### DISCUSSION TITLE
oop_class

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#14] [Announcements]
类 | LabVIEW 编程经验

https://lv.qizhen.xyz/oop_class

**Comment by Arita9527** [comment]
大神，为桌子属性建立的自定义数据类型，建完后都是跑到根目录下的吗？再自己拖到类里面？

**Reply by ruanqizhen** [reply]
也可以直接在类的新建菜单里选择建立”自定义数据类型“，我有时会直接从一个控件创建自定义数据类型，然后存储时候选择存储路径就行了。

**Comment by Arita9527** [comment]
把一个类拉到前面板和后面板，有什么区别呢？那种比较好？谢谢

**Reply by ruanqizhen** [reply]
在LabVIEW里面，不只是类，几乎任何一种数据，如果拖到前面板，就会添加一个控件；拖到后面板，就会添加一个常量。控件和常量的选择就完全取决于你需要哪一种了？数据不变的，一般用常量；如果数据从上层VI传进来的，需要用控件。

**Reply by StephenYang2022** [reply]
类的前面板控件，仅仅就是为了子vi接线端设置用？  这个正方体的控件，困扰了我很久。 
LabVIEW为何不引入没有前面板的控件，比如C#中的Timer定时器、串口控件等。

**Reply by ruanqizhen** [reply]
是的，是为了做子 VI 接线端。
这个设计可能并不是最优的。我估计是因为历史原因吧，有很多底层的设计，就算不合理也很难改变了。

**Comment by Arita9527** [comment]
因为这个 VI 要处理所以类型的家具---》因为这个 VI 要处理所有类型的家具

**Reply by ruanqizhen** [reply]
已修改

**Comment by Arita9527** [comment]
为什么 setup_funitures.vi 中furniture in到out连线是小箭头？？

**Reply by ruanqizhen** [reply]
在类的设置里面，可以选择这个类数据线的颜色和样式。这里是为了方便区分不同的类的数据，故意为每个类选择差距较大的数据线样式。我会在文档里再加个截图说明一下。

**Comment by Arita9527** [comment]
在程序调用了 assemble.vi 的地方双击这个子 VI，LabVIEW 不会像对待普通 VI 那样立刻打开子 VI，而是会列出所有类中的同名 VI，询问用户需要看哪一个。
？？？
为什么我看不到他说的提示图片？难道因为我都是用动态建立的？

**Reply by ruanqizhen** [reply]
需要有多个子类都重写了同一个方法，才会看到这个选择。

**Reply by Arita9527** [reply]
建立子类的时候，新建VI动态分配模板就是重写了吧？我名字都是写的assemble.vi
都没有出现你的那个提示框，奇怪了。

**Reply by ruanqizhen** [reply]
参数类型，接线板也要保持一致。你试试看用”用于重写的VI“这个选项新建VI行不行，这个选项可以帮你把必要的参数都设置好。
另外，再到类的继承关系里（下图）看看继承关系设置的对不对。

**Comment by Arita9527** [comment]
在工程浏览器中，如何一眼就看出，哪个类是继承哪个呢？有标记吗

**Reply by ruanqizhen** [reply]
最直接的方式是打开LabVIEW的菜单“查看->LabVIEW类层次结构”：
![](https://github.com/ruanqizhen/labview_book/raw/main/docs/images_2/image62.png)

**Comment by Yang-Bo-Wen** [comment]
能通过调用节点来调用实例化类的方法么

**Reply by ruanqizhen** [reply]
不能。我猜测原因可能是：调用节点每次也只能调用一个方法，没什么优势；而直接调用子VI的方式，程序可读性更好，也可以直接打开被调用VI，方便进行调试。相比起来调用节点似乎只有缺点，不添加这个功能也好。

**Comment by LZLAlexLZL** [comment]
老师，想问一下在chair子类下面的construct.vi中，对象输出和按名称捆绑之间无法正常连线，help显示是由于类的私有数据为空，无法捆绑或解除捆绑该类，请问这个问题应该如何解决

**Reply by ruanqizhen** [reply]
根据错误提示看，似乎是因为你还没有给类添加过任何任何数据，所以不能使用。先给类添加一些数据试试。

**Reply by ruanqizhen** [reply]
或者你想设置的是父类中的数据，就需要通过调用父类的方法来设置父类中的数据。

**Comment by StephenYang2022** [comment]
“最右边一部分，把两把椅子和一张桌子防止在了一个数组中”  发现了一个错误，应该是——“放置”。

老师，请问，所谓“构造方法”与类的一般方法有什么区别吗？ 如果没区别，为啥你叫它构造方法？

**Reply by ruanqizhen** [reply]
错字已修改，多谢指正！

构造函数（方法）与一般方法没有本质区别，只不过它指的是一类特定功能的方法：专用于在创建对象时初始化对象，也就是为对象成员变量（数据）赋初始值。

**Comment by Freddd13** [comment]
老师请问怎么实现在不同位置访问同一个类实例呢？最开始用的功能全局变量，结果发现类实例似乎也是传值的，但又不知道怎么使用引用，求教= =

**Reply by ruanqizhen** [reply]
如果可能的话，尽量使用值传递，这样比较符合 LabVIEW 编程习惯。无论想在哪使用类实例，把数据线连过去就行。
如果必须传引用，可以参考 <https://lv.qizhen.xyz/pattern_pass_by_ref> 中提到的几个方法。在后面还有一个用[双向链表](https://lv.qizhen.xyz/oop_use_cases#%E5%8F%8C%E5%90%91%E9%93%BE%E8%A1%A8-doubly-linked-list-%E6%88%96-double-linked-list)来演示的具体示例。

**Reply by Freddd13** [reply]
现在需要添加类实例的项目vi非常大，如果不方便通过连线方式去连接类方法的话，就得用引用或者值传递；我的问题是：
1. 我给类实例右键创建了一个引用，但这个引用不能连接到元素同址的引用输入端口。这种情况下怎么去解引用呢？
2. 如果使用值传递，那如果在不同地方去修改这个类，不能连线的情况下我就得保证各处使用的类值相同。我现在就是用了个功能全局变量，输入Get Set Create的枚举来保证值的一致，类似单例模式。这目前结果上是正确的，但是每次修改类属性都要Set一下，感觉很麻烦。请问在不能直接连线的前提下最佳实践是什么的呢？

**Reply by ruanqizhen** [reply]
1. 用鼠标右键创建的引用一般是“[控件的引用](https://lv.qizhen.xyz/vi_server_for_ui#%E5%BC%95%E7%94%A8)”，只有在复杂界面操做时会用到。元素同址结构少那个引用是专门处理“数据的引用”的。关于数据的引用创建和使用，可以参考：[数据引用节点](https://lv.qizhen.xyz/pattern_pass_by_ref#数据引用节点)
2. 如果不想使用数据线，那就只能使用全局变量；使用全局变量 set 操作就不可避免。不过最理想状态还是尽可能减小单个 VI 的复杂度，多划分子 VI。如果数据线太多，可以考虑把一些功能相关的数据捆绑成簇或者封装成类。

**Reply by Freddd13** [reply]
感谢老师回复，还是有点疑惑…
1. 我尝试过您说的新建数据值引用，但我的类实例不能连接到它的输入端，好像是报错多态vi不支持此类型。这个类是我拖到前面板的，不知道是不是这里有问题？

2.我现在不是在一个顺序结构去用类的，而是在几个循环/事件监听中之间都要使用。这种情况下使用值传递似乎也避免不了set操作？不然怎么把修改后的最新状态值同步回去呢？
(vi分割方面，由于是好几个人加了几年的代码，比较庞大，实在是难以修改，封装个类也是为了避免再把新功能的属性和方法耦合进去)

**Reply by Freddd13** [reply]
> 感谢老师回复，还是有点疑惑…
> 1. 我尝试过您说的新建数据值引用，但我的类实例不能连接到它的输入端，好像是报错多态vi不支持此类型。这个类是我拖到前面板的，不知道是不是这里有问题？
> 
> 2.我现在不是在一个顺序结构去用类的，而是在几个循环/事件监听中之间都要使用。这种情况下使用值传递，就算连线似乎也避免不了set操作？不然怎么把修改后的最新状态值同步回去呢？
> (vi分割方面，由于是好几个人加了几年的代码，比较庞大，实在是难以修改，封装个类也是为了避免再把新功能的属性和方法耦合进去)

**Reply by ruanqizhen** [reply]
1. 我不是太确定你遇到的问题，我也很久没用过 LabVIEW 了，可能是某些方面有限制。我在这里的示例 VI 有用到应用，你可以参考一下看看和你的用法是否相同 <https://lv.qizhen.xyz/oop_use_cases#idoublynode-%E6%8E%A5%E5%8F%A3%E5%92%8C-doublynode-%E7%B1%BB>
2. 如果实在几个并行的结构内使用同一个类实例，并且还要改变类中的数据，那么只能使用传引用，可以有数据线的，也可以是全局变量，全局变量避免不了set操作。

**Comment by Freddd13** [comment]
您的那个示例我看过了，我感觉用法上和我的应该是差不多的，我的labview可能版本略低打不开您那个示例，但里面接引用的方法在我自己这个类上暂时是没尝试成功。 第二个我明白了，现阶段还是打算先用全局变量了。最后感谢老师的耐心回复！
在 2024年9月24日+0800 06:02，ruanqizhen ***@***.***>，写道：

**Comment by luck-zhong** [comment]
"这里的实例是由 Child 类生成的，不论它在程序中使用哪个祖先类的数据类型表示它，它都始终还是一个 Child 类实例，所以程序调用它的 dynamic.vi，运行的一定还是 Child 类中的那个 dynamic.vi。"实例由哪个类创建是根据传入的类来决定的，而跟你使用的数据类型无关，可以这么理解吗？那转换成父类的数据类型有什么作用吗？

**Reply by ruanqizhen** [reply]
是的，在创建实例（对象）时，其类型就已经确定了。之后，即使使用父类类型传递对象，也不会改变其原始类型。
在 LabVIEW 没有接口之前，常将子类转换成父类，以便于上层程序调用。这使得调用者可以使用更通用的数据类型，调用一些不同类型的数据。比如，通过“动物类”调用多个动物对象，而无需关心它们具体是猫对象还是狗对象。
然而，有了接口之后，所有调用都应通过接口进行，而不是通过父类。现在，父类的主要作用是实现一些所有子类共同的代码，供子类继承，从而减少代码重复。
一般情况下，不再需要在程序中将对象类型转换成父类。

**Reply by luck-zhong** [reply]
您所说的接口指的是？

**Reply by ruanqizhen** [reply]
再下面一个章节就有详细介绍接口
